chore(web-analytics): refactoring EventsHeatmap to scale

jabahamondes opened the pull request:
## Problem

Currently, events heatmap is too coupled with the use case of active hours feature. This is the first refactor that decouples the logic into props. 

## Changes

Create a new component that wraps up eventheatmap component passing the desired props

## Does this work well for both Cloud and self-hosted?

Yes

## How did you test this code?

Updating tests and manually.

Commit:
feat(web-analytics): add loading state to EventsHeatMap component

- Introduced loading state handling using InsightLoadingState and InsightsWrapper.
- Updated useValues to include responseLoading and queryId for better loading management.

Commit:
feat(web-analytics): add Active Hours tab and update priority mapping

- Introduced ACTIVE_HOURS enum to TileId and ActiveHoursTab enum.
- Updated loadPriorityMap to include ACTIVE_HOURS with adjusted priorities for existing tiles.
- Added setActiveHoursTab action and state management in webAnalyticsLogic.
- Integrated Active Hours tab into the web analytics logic and UI.

Commit:
fix(web-analytics): update web active hours heatmap query for unique session events

- Refactored the query to use unique session events for aggregating hours and days.
- Adjusted the subquery to group by session_id and ensure correct timestamp handling.

Commit:
Moving InsightsWrapper to own place

Commit:
Merge branch 'master' into feat/making-query-handle-unique-users-for-activity-hours

Commit:
Removing extra space

jabahamondes committed:
Merge branch 'master' into feat/making-query-handle-unique-users-for-activity-hours

Commit:
Update tests

Commit:
Add active hours documentation and enhance EventsHeatMap logic

- Introduced detailed documentation for the active hours feature in webAnalyticsLogic.tsx, including usage notes and recommendations for time range selection.
- Updated EventsHeatMap component to incorporate week start day logic, improving the accuracy of the heatmap display.
- Refactored data processing functions to handle new aggregation structures and ensure proper rendering of heatmap cells with updated tooltip functionality.

Commit:
Merge branch 'master' into feat/improving-activty-hours

Commit:
Refactor heatmap query structure and components

- Introduced new HeatMapQueryResponse and HeatMapStructuredResult types to replace the previous WebActiveHoursHeatMap structures.
- Updated EventsHeatMap component to utilize WebActiveHoursHeatmap for rendering, enhancing the heatmap display logic.
- Removed deprecated WebActiveHoursHeatMap related code and tests, streamlining the overall query handling.
- Added utility functions for tooltip generation and font size adjustments based on table width.
- Improved data processing logic to accommodate new data structures and ensure accurate rendering of heatmap cells.

Commit:
Merge branch 'master' into chore/refactoring-events-heatmap-to-scale-to-more-queries

Commit:
fix(EventsHeatMap): simplify font size calculation logic

Refactored the font size setting in the EventsHeatMap component to directly call the thresholdFontSize function, removing unnecessary intermediate variable. This change streamlines the code and improves readability.

Commit:
fix(heatmap): Correct background color logic for min/max values in HeatMapCell
- Adjusted the order of conditions to ensure proper saturation for min and max values.

Commit:
- Changed matrix type from object to a 2D array for better performance and simplicity.
- Updated columns and rows aggregations to use arrays instead of objects, enhancing consistency across the data processing functions.

pauldambra commented:
ah, nice, was going to ask if I could use this to show session recordings too, so I'm happy about this refactor :)

Commit:
naming mixed fixing

Commit:
refactor(heatmap): Rename WebActiveHoursHeatMapQuery to HeatMapQuery
- Updated schema and query references to use HeatMapQuery instead of WebActiveHoursHeatMapQuery for consistency.
- Introduced new HeatMapQuery definition in schema.json and updated related utility functions and components accordingly.
- Adjusted all instances in the codebase to reflect the new naming convention, ensuring seamless integration with existing features.

Commit:
handling negative module

Commit:
Merge branch 'master' into chore/refactoring-events-heatmap-to-scale-to-more-queries

lricoy commented on frontend/src/queries/Query/Query.tsx:
```diff
@@ -175,8 +175,8 @@ export function Query<Q extends Node>(props: QueryProps<Q>): JSX.Element | null
         component = <HogDebug query={query} setQuery={setQuery as (query: any) => void} queryKey={String(uniqueKey)} />
-    } else if (isWebActiveHoursHeatMapQuery(query)) {
-        component = <EventsHeatMap query={query} context={queryContext} cachedResults={props.cachedResults} />
+    } else if (isHeatMapQuery(query)) {
```
I would suggest to keep `ActiveHoursHeatmap` or `EventsHeatMapQuery` because we do have heatmaps as well: https://posthog.com/docs/toolbar/heatmaps

lricoy commented on frontend/src/scenes/web-analytics/webAnalyticsLogic.tsx:
```diff
@@ -1843,9 +1843,16 @@ export const webAnalyticsLogic = kea<webAnalyticsLogicType>([
                                       canOpenModal: true,
                                       query: {
-                                          kind: NodeKind.WebActiveHoursHeatMapQuery,
+                                          kind: NodeKind.HeatMapQuery,
+                                          source: {
+                                              kind: NodeKind.EventsNode,
+                                              event: '$pageview',
+                                              name: '$pageview',
+                                              math: BaseMathType.UniqueUsers,
+                                          },
                                           properties: webAnalyticsFilters,
                                           dateRange,
+                                          conversionGoal,
```
Is this changing anything for your query or was it just for the type checking?

The `conversionGoal` is used when selecting a event as the conversion goal but it must be handled by the query. The `source` is probably going to be ignored as we're using a custom query but I may be wronged there.

<img width="886" alt="image" src="https://github.com/user-attachments/assets/9730bd71-6234-4bc8-8f02-87a98cb0e94c" />

Commit:
renaming

Commit:
fixing tests

github-actions[bot] committed:
Update UI snapshots for `chromium` (5)

Commit:
Merge branch 'master' into chore/refactoring-events-heatmap-to-scale-to-more-queries

Commit:
Merge branch 'chore/refactoring-events-heatmap-to-scale-to-more-queries' of github.com:PostHog/posthog into chore/refactoring-events-heatmap-to-scale-to-more-queries

github-actions[bot] committed:
Update UI snapshots for `chromium` (5)
